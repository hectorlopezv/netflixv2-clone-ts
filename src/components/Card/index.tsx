/* eslint-disable react/jsx-pascal-case */
import React, { useState, useContext, createContext } from 'react';
import { 
    Container,
    Title, 
    Group ,
    SubTitle ,
    Text ,
    Feature,
    Feature_Title ,
    Feature_Text ,
    Feature_Close ,  
    maturity as Maturity,
    Content,
    Meta,
    Item ,
    Image, 
    Entities,
    } from './styles/Card';


export const FeatureContext = createContext<{[prop:string]:any}>({});

export interface CardProps {

}

const Card: React.FC<CardProps> = ({ children, ...props }) => {
    //when we click card we need a state to show or not Feautures
    ///we want to show only one feaute in one group

    const [showFeature, setshowFeature] = useState(false);
    const [itemFeature, setitemFeature] = useState({});
    return (  
        <FeatureContext.Provider value={{showFeature, setshowFeature, itemFeature, setitemFeature}} >
            <Container {...props}>{children}</Container>
        </FeatureContext.Provider>

    );
}
export default Card;


export interface Card_GroupProps {
    
}
 
export const Card_Group: React.FC<Card_GroupProps> = ({ children, ...props }) => {
    return (  
        <Group {...props}>{children}</Group>
    );
}
 

export interface Card_TitleProps {
    
}
 
export const Card_Title: React.FC<Card_TitleProps> = ({ children, ...props }) => {
    return (  <Title {...props}>{children}</Title>);
}
 
export interface Card_SubTitleProps {
    
}
 
export const Card_SubTitle: React.FC<Card_SubTitleProps> = ({ children, ...props }) => {
    return ( <SubTitle {...props}>{children}</SubTitle> );
}
 
export interface Card_TextProps {
    
}
 
export const Card_Text: React.FC<Card_TextProps> = ({ children, ...props }) => {
    return ( <Text {...props}>{children}</Text> );
}
 

export interface Card_MetaProps {
    
}
 
export const Card_Meta: React.FC<Card_MetaProps> =  ({ children, ...props }) => {
    //description of the series
    //holder div for the info
    return ( <Meta {...props}>{children}</Meta> );
}
 
export interface Card_ItemProps {
    item: any;
}
 
export const Card_Item: React.FC<Card_ItemProps> =  ({ item, children, ...props }) => {
    //we we use the shared state
    const {setshowFeature, setitemFeature} = useContext(FeatureContext);
    return ( <Item onClick={() => {
        setitemFeature(item);
        setshowFeature(true);
    }}
        {...props}>{children}</Item> );
}
 
export interface Card_ImageProps {
    src: string;
}
 
export const Card_Image: React.FC<Card_ImageProps> = ({...props}) => {
    return ( <Image {...props} /> );
}


export interface Card_EntitiesProps {
    
}
 
export const Card_Entities: React.FC<Card_EntitiesProps> = ({children, ...props}) => {
    return ( <Entities {...props}>{children}</Entities> );
}
 

export interface Card_FeatureProps {
    category: any;
}

//the modal with react PORTAL
export const Card_Feature: React.FC<Card_FeatureProps> = ({category, children, ...props}) => {
    const {showFeature, setshowFeature, itemFeature, setitemFeature} = useContext(FeatureContext);
    return showFeature ? (
        <Feature 
            src={`/images/${category}/${itemFeature.genre}/${itemFeature.slug}/large.jpg`} 
            {...props}
        >
            <Content>
                <Feature_Title>{itemFeature.title}</Feature_Title>
                <Feature_Text>{itemFeature.description}</Feature_Text>
                
                <Feature_Close onClick={() => setshowFeature(false)}>
                    <img src={"/images/icons/close.png"} alt="Close"/>
                </Feature_Close>
                {children}
            </Content>

            <Group margin="30px 0" flexDireciton="row" alignItems="center">
                <Maturity rating={itemFeature.maturity}>
                    {itemFeature.maturity < 12 ? 'PG' : itemFeature.maturity}
                </Maturity>
                <Feature_Text fontWeight="bold">
                    {itemFeature.genre.charAt(0).toUpperCase() + itemFeature.genre.slice(1)}
                </Feature_Text>
            </Group>
        </Feature> 
    ): null;
}
 
