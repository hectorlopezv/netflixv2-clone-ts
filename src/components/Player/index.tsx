import React, {useState, useContext, createContext} from 'react';
//here is the big MODAL using React Portal
import ReactDom from 'react-dom';
import {Container, Button, Overlay, Inner, Close} from './styles/Player';

export interface PlayerProps {
    
}
const PlayerContext = createContext<{[prop: string]: any}>({});
const Player: React.FC<PlayerProps> = ({children, ...props}) => {
    const [showPlayer, setshowPlayer] = useState(false);

    return (  
        <PlayerContext.Provider value={{showPlayer, setshowPlayer}}>
            <Container {...props}>{children}</Container>
        </PlayerContext.Provider> 
    );
}

export default Player;


export interface Player_VideoProps {
    src: string;
}
 
export const Player_Video: React.FC<Player_VideoProps> = ({src, ...props}) => {
    const { showPlayer, setshowPlayer } = useContext(PlayerContext);
    return showPlayer ? ReactDom.createPortal(
    <Overlay data-testid="player" onClick={() => setshowPlayer(false)  }>
        <Inner>
            <video id="netflix-player" controls>
                <source src={src} type="video/mp4" />
            </video>
            <Close />
        </Inner>
    </Overlay>, document.body): null;
}


export interface Player_ButtonProps {
    
}
 
export const Player_Button: React.FC<Player_ButtonProps> = () => {
    const { showPlayer, setshowPlayer } = useContext(PlayerContext);

    return <Button onClick={() => setshowPlayer( (prevstate: boolean) => !prevstate )}>
        Play
    </Button>
}
 

 