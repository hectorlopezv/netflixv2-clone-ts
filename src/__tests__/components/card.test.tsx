/* eslint-disable react/jsx-pascal-case */
import React from 'react';
import { render, fireEvent, getByTestId, getByAltText } from '@testing-library/react';


import Card, 
{Card_Text,
    Card_Feature,
    Card_Group, Card_Title, Card_Entities, Card_Item, Card_Image, Card_Meta, Card_SubTitle
} from '../../components/Card';

import Player, 
{
    Player_Button, Player_Video
} from '../../components/Player';



const category = 'series';
const slideRows = [
  {
    title: 'Documentaries',
    data: [
      {
        genre: 'documentaries',
        maturity: '18',
        slug: 'tiger-king',
        description: 'Tiger King description',
        id: '07adcf19-517f-41b4-8f8b-04fee694bce1',
        title: 'Tiger King',
        docId: 'IcRxJBanzE7nmIp2GMGb',
      },
    ],
  },
  {
    title: 'Feel Good',
    data: [
      {
        title: 'Juno',
        genre: 'feel-good',
        description: 'Juno description',
        maturity: '0',
        id: 'a938b5a6-fe25-4a06-8050-353bc7146ccd',
        slug: 'juno',
        docId: '4JDgdf5vE0K3YrW9ZnbP',
      },
    ],
  },
];


describe('<Card /> Component', () => {

    it('<Card > Renders the card with populated data', () => {
        const {container, getByText, queryByText} =  render (
            <Card_Group>
                {slideRows.map((slideItem: any) => (
                    <Card key={`${category}-${slideItem.title.toLowerCase()}`}>{/* carc container*/}
                        <Card_Title>{slideItem.title}</Card_Title>
                        <Card_Entities>{/*Card Carrousel*/}
                            {/* Card Item*/}
                            {slideItem.data.map((item: any) => ( 
                                <Card_Item key={item.docId} item={item}>
                                    <Card_Image src={`/images/${category}/${item.genre}/${item.slug}/small.jpg`}/>
                                    <Card_Meta>
                                        <Card_SubTitle>{item.title}</Card_SubTitle>
                                        <Card_Text>{item.description}</Card_Text>
                                    </Card_Meta>
                                </Card_Item>
                            ))}
                        </Card_Entities>
                        <Card_Feature category={category}>
                            <Player>
                                <Player_Button />
                                <Player_Video src={"/videos/bunny.mp4"}/>
                            </Player>
                        </Card_Feature>

                    </Card> 
                ))}
            </Card_Group>
        );

        //asertions
        expect(getByText('Documentaries')).toBeTruthy();
        expect(getByText('Tiger King')).toBeTruthy();
        expect(getByText('Tiger King description')).toBeTruthy();

        expect(getByText('Feel Good')).toBeTruthy();
        expect(getByText('Juno')).toBeTruthy();
        expect(getByText('Juno description')).toBeTruthy();

        expect(container.firstChild).toMatchSnapshot();


    });


    it('renders <Card > testing toogling  setFeature and showFeature', () => {
        const {container, getByText, queryByText, getByAltText, queryByTestId, getByTestId} =  render (
            <Card_Group>
                {slideRows.map((slideItem: any) => (
                    <Card key={`${category}-${slideItem.title.toLowerCase()}`}>{/* carc container*/}
                        <Card_Title>{slideItem.title}</Card_Title>
                        <Card_Entities>{/*Card Carrousel*/}
                            {/* Card Item*/}
                            {slideItem.data.map((item: any) => ( 
                                <Card_Item key={item.docId} item={item} data-testid={`${item.slug}-item-feature`}>
                                    <Card_Image src={`/images/${category}/${item.genre}/${item.slug}/small.jpg`}/>
                                    <Card_Meta>
                                        <Card_SubTitle>{item.title}</Card_SubTitle>
                                        <Card_Text>{item.description}</Card_Text>
                                    </Card_Meta>
                                </Card_Item>
                            ))}
                        </Card_Entities>
                        <Card_Feature category={category}>
                            <Player>
                                <Player_Button />
                                <Player_Video src={"/videos/bunny.mp4"}/>
                            </Player>
                        </Card_Feature>

                    </Card> 
                ))}
            </Card_Group>
        );

        //asertions
        expect(queryByText('18')).toBeFalsy();
        fireEvent.click(getByTestId('tiger-king-item-feature')!);
        expect(queryByText('18')).toBeTruthy();

        //click close button
        fireEvent.click(getByAltText('Close')!);
        expect(queryByText('18')).toBeFalsy();

        expect(queryByText('PG')).toBeFalsy();
        
        fireEvent.click(getByTestId('juno-item-feature')!);
        expect(queryByText('PG')).toBeTruthy();

        fireEvent.click(getByAltText('Close')!);
        expect(queryByText('PG')).toBeFalsy();

        expect(container.firstChild).toMatchSnapshot();
        
    });
});