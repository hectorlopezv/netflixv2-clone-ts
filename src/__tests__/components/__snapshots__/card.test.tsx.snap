// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> Component <Card > Renders 1`] = `
<div
  class="Card__Group-sc-56ol5s-2 jsoZXk"
>
  <div
    class="Card__Container-sc-56ol5s-1 UryZj"
  >
    <p
      class="Card__Title-sc-56ol5s-0 hGznPR"
    >
      Documentaries
    </p>
    <div
      class="Card__Entities-sc-56ol5s-13 jGXcjx"
    >
      <div
        class="Card__Item-sc-56ol5s-11 fWjLSO"
      >
        <img
          class="Card__Image-sc-56ol5s-12 ikuVEa"
          src="/images/series/documentaries/tiger-king/small.jpg"
        />
        <div
          class="Card__Meta-sc-56ol5s-10 kWMUtp"
        >
          <p
            class="Card__SubTitle-sc-56ol5s-3 fTmGSZ"
          >
            Tiger King
          </p>
          <p
            class="Card__Text-sc-56ol5s-4 dbAbDs"
          >
            Tiger King description
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    class="Card__Container-sc-56ol5s-1 UryZj"
  >
    <p
      class="Card__Title-sc-56ol5s-0 hGznPR"
    >
      Feel Good
    </p>
    <div
      class="Card__Entities-sc-56ol5s-13 jGXcjx"
    >
      <div
        class="Card__Item-sc-56ol5s-11 fWjLSO"
      >
        <img
          class="Card__Image-sc-56ol5s-12 ikuVEa"
          src="/images/series/feel-good/juno/small.jpg"
        />
        <div
          class="Card__Meta-sc-56ol5s-10 kWMUtp"
        >
          <p
            class="Card__SubTitle-sc-56ol5s-3 fTmGSZ"
          >
            Juno
          </p>
          <p
            class="Card__Text-sc-56ol5s-4 dbAbDs"
          >
            Juno description
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Card /> Component <Card > Renders the card with populated data 1`] = `
<div
  class="Card__Group-sc-56ol5s-2 jsoZXk"
>
  <div
    class="Card__Container-sc-56ol5s-1 UryZj"
  >
    <p
      class="Card__Title-sc-56ol5s-0 hGznPR"
    >
      Documentaries
    </p>
    <div
      class="Card__Entities-sc-56ol5s-13 jGXcjx"
    >
      <div
        class="Card__Item-sc-56ol5s-11 fWjLSO"
      >
        <img
          class="Card__Image-sc-56ol5s-12 ikuVEa"
          src="/images/series/documentaries/tiger-king/small.jpg"
        />
        <div
          class="Card__Meta-sc-56ol5s-10 kWMUtp"
        >
          <p
            class="Card__SubTitle-sc-56ol5s-3 fTmGSZ"
          >
            Tiger King
          </p>
          <p
            class="Card__Text-sc-56ol5s-4 dbAbDs"
          >
            Tiger King description
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    class="Card__Container-sc-56ol5s-1 UryZj"
  >
    <p
      class="Card__Title-sc-56ol5s-0 hGznPR"
    >
      Feel Good
    </p>
    <div
      class="Card__Entities-sc-56ol5s-13 jGXcjx"
    >
      <div
        class="Card__Item-sc-56ol5s-11 fWjLSO"
      >
        <img
          class="Card__Image-sc-56ol5s-12 ikuVEa"
          src="/images/series/feel-good/juno/small.jpg"
        />
        <div
          class="Card__Meta-sc-56ol5s-10 kWMUtp"
        >
          <p
            class="Card__SubTitle-sc-56ol5s-3 fTmGSZ"
          >
            Juno
          </p>
          <p
            class="Card__Text-sc-56ol5s-4 dbAbDs"
          >
            Juno description
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Card /> Component <Card > Renders the card with populated data 2`] = `
<div
  class="Card__Group-sc-56ol5s-2 jsoZXk"
>
  <div
    class="Card__Container-sc-56ol5s-1 UryZj"
  >
    <p
      class="Card__Title-sc-56ol5s-0 hGznPR"
    >
      Documentaries
    </p>
    <div
      class="Card__Entities-sc-56ol5s-13 jGXcjx"
    >
      <div
        class="Card__Item-sc-56ol5s-11 fWjLSO"
      >
        <img
          class="Card__Image-sc-56ol5s-12 ikuVEa"
          src="/images/series/documentaries/tiger-king/small.jpg"
        />
        <div
          class="Card__Meta-sc-56ol5s-10 kWMUtp"
        >
          <p
            class="Card__SubTitle-sc-56ol5s-3 fTmGSZ"
          >
            Tiger King
          </p>
          <p
            class="Card__Text-sc-56ol5s-4 dbAbDs"
          >
            Tiger King description
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    class="Card__Container-sc-56ol5s-1 UryZj"
  >
    <p
      class="Card__Title-sc-56ol5s-0 hGznPR"
    >
      Feel Good
    </p>
    <div
      class="Card__Entities-sc-56ol5s-13 jGXcjx"
    >
      <div
        class="Card__Item-sc-56ol5s-11 fWjLSO"
      >
        <img
          class="Card__Image-sc-56ol5s-12 ikuVEa"
          src="/images/series/feel-good/juno/small.jpg"
        />
        <div
          class="Card__Meta-sc-56ol5s-10 kWMUtp"
        >
          <p
            class="Card__SubTitle-sc-56ol5s-3 fTmGSZ"
          >
            Juno
          </p>
          <p
            class="Card__Text-sc-56ol5s-4 dbAbDs"
          >
            Juno description
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
